### PR TITLE
Abstract experiment logic

### PR DESCRIPTION
**Pull request by jaredsilver**

### CODE REVIEW

File: assets/js/libs/experiment.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    this.chooseVariant = function() {
+      var savedVariant = window.localStorage.getItem(this.EXPERIMENT_NAME);
+      if(savedVariant) { return savedVariant }
+
+      // Use a CDF to select a variant based on weight.
+      var total = this.variants.reduce(function(total, variant) {
+        return total + variant.weight;
+      }, 0);
+      var random = Math.floor(Math.random() * (total + 1));
```
Comment (vincentvankrunkelsven) [inline assets/js/libs/experiment.js]: Why not just use `Math.random() * total`?

File: assets/js/libs/experiment.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      var selectedVariant = this.variants.find(function(variant) {
+        random -= variant.weight;
```
Comment (vincentvankrunkelsven) [inline assets/js/libs/experiment.js]: To me it's a bit weird to mutate a value like this, but I guess it works.

A different way to tackle this, is to create a list of the cumulative weights (e.g. `const sums = [50, 100]` and use `sums.findIndex(x => x < Math.random() * sums[sums.length - 1])` or something.

I think either is fine though, just a remark (you usually don't expect the callback in `find` to mutate a value.

File: assets/js/libs/experiment.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+(function(){
+  window.Experiment = function(experimentName, experimentClass) {
+    this.EXPERIMENT_NAME = experimentName;
+    this.variants = [];
+    this.experimentClass = experimentClass;
+
+    this.addVariant = function(name, weight) {
+      if(weight < 1) { return alert('Snowplow requires weights to be integers.'); }
```
Comment (vincentvankrunkelsven) [inline assets/js/libs/experiment.js]: Is this here to avoid people using proportions (e.g. `0.5`) instead of weights?

**Comment by jaredsilver** [comment]
:+1: Will address code review in #411.

**Pull request merged**